Cython/Compiler/Nodes.py

```diff
@@ -4731,8 +4725,152 @@ def generate_execution_code(self, code):
+              # Let Python do the base types compatibility checking.
+              trial_type = code.funcstate.allocate_temp(PyrexTypes.py_object_type, True)
+              code.putln("%s = PyType_Type.tp_new(&PyType_Type, %s, NULL);" % (
+                  trial_type, self.type_init_args.result()))
+              code.putln(code.error_goto_if_null(trial_type, self.pos))
+              code.put_gotref(trial_type)
+              code.putln("if (((PyTypeObject*) %s)->tp_base != %s) {" % (
+                  trial_type, first_base))
+              code.putln("PyErr_Format(PyExc_TypeError, \"best base '%s' must be equal to first base '%s'\",")
+              code.putln("             ((PyTypeObject*) %s)->tp_base->tp_name, %s->tp_name);" % (
+                         trial_type, first_base))
+              code.putln(code.error_goto(self.pos))
+              code.putln("}")
+              code.funcstate.release_temp(trial_type)
+              code.put_incref(bases, PyrexTypes.py_object_type)
+              code.put_giveref(bases)
+              code.putln("%s.tp_bases = %s;" % (self.entry.type.typeobj_cname, bases))
+              code.put_decref_clear(trial_type, PyrexTypes.py_object_type)
+              self.type_init_args.generate_disposal_code(code)
+              self.type_init_args.free_temps(code)
+
+            self.generate_type_ready_code(self.entry, code, True)
+
+    # Also called from ModuleNode for early init types.
+    @staticmethod
+    def generate_type_ready_code(entry, code, heap_type_bases=False):
+        # Generate a call to PyType_Ready for an extension
+        # type defined in this module.
+        type = entry.type
+        typeobj_cname = type.typeobj_cname
+        scope = type.scope
+        if scope: # could be None if there was an error
+            if entry.visibility != 'extern':
+                for slot in TypeSlots.slot_table:
+                    slot.generate_dynamic_init_code(scope, code)
+                if heap_type_bases:
+                  # As of https://bugs.python.org/issue22079
+                  # PyType_Ready enforces that all bases of a non-heap type
+                  # are non-heap.  We know this is the case for the solid base,
+                  # but other bases may be heap allocated and are kept alive
+                  # though the bases reference.
+                  # Other than this check, this flag is unused in this method.
+                  code.putln("#if PY_VERSION_HEX >= 0x03050000")
+                  code.putln("%s.tp_flags |= Py_TPFLAGS_HEAPTYPE;" % typeobj_cname)
+                  code.putln("#endif")
+                code.putln(
+                    "if (PyType_Ready(&%s) < 0) %s" % (
+                        typeobj_cname,
+                        code.error_goto(entry.pos)))
+                if heap_type_bases:
+                  code.putln("#if PY_VERSION_HEX >= 0x03050000")
+                  code.putln("%s.tp_flags &= ~Py_TPFLAGS_HEAPTYPE;" % typeobj_cname)
```

Comment:
Hmm. Eventually, we will have to switch to heap-types anyway to properly support PEP 489, submodules, reloading, etc. I understand that this doesn't have to happen *right* now, but you would probably agree that this flag unsetting seems like a fairly crude hack.

Reply:
Yes, this is an ugly hack (and one of my hesitations for the whole PR) but I couldn't figure out a way around it without making everything heap allocated (a much bigger change, though, as you said, we may need to go there).